networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/AttestationSubnetService.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.eth2;
+
+import java.util.Collections;
+import java.util.function.Consumer;
+import tech.pegasys.artemis.util.events.Subscribers;
+
+/**
+ * Service tracks long term attestation subnet subscriptions and notifies subscribers on their
+ * changes
+ */
+public class AttestationSubnetService {
+  private final Subscribers<Consumer<Iterable<Integer>>> subscribers = Subscribers.create(true);
+  private volatile Iterable<Integer> currentSubscriptions = Collections.emptyList();
+
+  public void updateSubscriptions(final Iterable<Integer> subnetIndices) {
+    subscribers.deliver(Consumer::accept, subnetIndices);
+    currentSubscriptions = subnetIndices;
+  }
+
+  public Iterable<Integer> getCurrentAttestatoinSubnetSubscriptions() {
+    return currentSubscriptions;
+  }
+
+  public long subscribeToUpdates(Consumer<Iterable<Integer>> observer) {
+    observer.accept(getCurrentAttestatoinSubnetSubscriptions());
```

Comment:
Each new subscriber receives the latest data. This method is not thread-safe but it's ok for now since we are subscribing on initialization step only. If no one minds I would add a reactive-streams dependency (I think Reactor it the best fit) and make this subscription stuff based on `Publisher` to not reinvent the wheel.

Reply:
I'm not sure I see the threading issue here.  The `Subscribers` class is specifically intended for this kind of subscriber pattern and is thread safe.  `current subscriptions` is volatile so the only risk is if `updateSubscriptions` is passed something that's also modified externally (which is a problem even ignoring threads).

I guess it's also possible that multiple threads make simultaneous calls to `updateSubscriptions` but there's no way to replace the whole list concurrently and not wind up losing one of the changes so it will have to be the upstream caller that handles that.

Reply:
@ajsutton 
Yep the `Subscribers` is thread-safe, but the `AttestationSubnetService` is not. 
What I'd like to have is the guarantee that:
- any `subscriber` added at any moment from any thread would receive the latest item. 
- items are delivered in the right order
- no duplicate items 

The current implementation may violate all of the above restrictions. You just need to invoke `subscribeToUpdates` and `updateSubscriptions` from different threads. (`updateSubscriptions` may always be called from a single thread)

I tried to implement it (that `Publisher` class) but it appeared a bit tricky as I don't want the listener callbacks to be invoked under the lock since it is deadlock-prone.  So just doing both methods `synchronized` is not the case. 

That's why I though that using Reactive library might be handy even for these primitive cases just to not reinvent a wheel.

Reply:
Ah I see. The only reason we need that construct is because the list of subscriptions and the sequence number is being maintained in two separate places.  Given that the `PING` response needs to be able to provide the same sequence number, wouldn't it make more sense to maintain the sequence number in `AttestationSubnetService`? Then the listeners aren't required at all - both metadata and ping response handlers can just request the latest value and use it.

Reply:
The reason I wanted to separate `AttestationSubnetService` is that other services might be interested in the value updates. E.g. at the moment the `discoveryNetwork` also needs to be notified on updates. The `seq_number` relates only to Eth2 RPC, so I thought it makes sense to manage it in a more RPC specific class (`MetadataMessageFactory`). 

For serving the `PING` request I was going to use the same `MetadataMessageFactory` (though may be rename it appropriately)

Anyways I've made `AttestationSubnetService` thread-safe so we may safely stick to any option

Reply:
Finally came up with a thread safe `ObservableValue` class without a dedicated thread. 
The only drawback of `ObservableValue` is that upon subscribe it calls the update callback under the lock. But this should be safe enough. 
On updates subscribers callbacks are invoked without locking.